src/requirementslib/models/requirements.py

```diff
@@ -822,7 +822,7 @@ def as_line(self, sources=None):
 
     @property
     def constraint_line(self):
-        return self.as_line()
+        return self.as_line() + ("[" + ",".join([extra for extra in self.extras]) + "]" if self.is_vcs else "")
```

Comment:
You probably want to use `extras_as_pip` => https://github.com/sarugaku/requirementslib/blob/master/src/requirementslib/models/requirements.py#L679 (can you sort them while you're making this change? just toss a `sorted` on there should be good enough

Reply:
I'm not sure if this is the right solution because it would put the extras after specifiers, markers, and hashes as formatted by `as_line()`. The better solution is probably to have `as_line()` take an optional `force_extras` parameter:

```python
    def as_line(self, sources=None, *, force_extras=False):
        line = "{0}{1}{2}{3}{4}".format(
            self.req.line_part,
            self.extras_as_pip if force_extras or not self.is_vcs else "",
            self.specifiers if self.specifiers else "",
            self.markers_as_pip,
            self.hashes_as_pip,
        )
```

What do you think?

Reply:
(we will probably wind up just always including extras on vcs requirements, I can't really imagine not wanting them no matter how we output them, but this is a good fix to get us there)